GVFS/GVFS/CommandLine/UpgradeVerb.cs

```diff
@@ -0,0 +1,384 @@
+
+        protected override string VerbName
+        {
+            get { return UpgradeVerbName; }
+        }
+        
+        public override void Execute()
+        {
+            ReturnCode exitCode = ReturnCode.Success;
+            if (!this.TryInitializeUpgrader() || !this.TryRunProductUpgrade())
+            {
+                exitCode = ReturnCode.GenericError;
+                this.ReportErrorAndExit(this.tracer, exitCode, string.Empty);
+            }
+        }
+
+        private bool TryInitializeUpgrader()
+        {
+            OperatingSystem os_info = Environment.OSVersion;
+
+            if (os_info.Platform == PlatformID.Win32NT)
+            {
+                if (this.upgrader == null)
+                {
+                    JsonTracer jsonTracer = new JsonTracer(GVFSConstants.GVFSEtwProviderName, "UpgradeVerb");
+                    string logFilePath = GVFSEnlistment.GetNewGVFSLogFileName(
+                        ProductUpgrader.GetLogDirectoryPath(),
+                        GVFSConstants.LogFileTypes.UpgradeVerb);
+                    jsonTracer.AddLogFileEventListener(logFilePath, EventLevel.Informational, Keywords.Any);
+
+                    this.tracer = jsonTracer;
+                    this.prerunChecker = new InstallerPreRunChecker(this.tracer, this.Confirmed ? GVFSConstants.UpgradeVerbMessages.GVFSUpgradeConfirm : GVFSConstants.UpgradeVerbMessages.GVFSUpgrade);
+                    this.upgrader = new ProductUpgrader(ProcessHelper.GetCurrentProcessVersion(), this.tracer);
+                }
+
+                return true;
+            }
+            else
+            {
+                this.ReportInfoToConsole($"ERROR: {GVFSConstants.UpgradeVerbMessages.GVFSUpgrade} in only supported on Microsoft Windows Operating System.");
+                return false;
+            }
+        }
+
+        private bool TryRunProductUpgrade()
+        {
+            string errorOutputFormat = Environment.NewLine + "ERROR: {0}";
+            string error = null;
+            string cannotInstallReason = null;
+            Version newestVersion = null;
+            ProductUpgrader.RingType ring = ProductUpgrader.RingType.Invalid;
+
+            bool isInstallable = this.TryCheckUpgradeInstallable(out cannotInstallReason);
+            if (this.Confirmed && !isInstallable)
+            {
+                this.ReportInfoToConsole($"Cannot upgrade GVFS on this machine.");
+                this.Output.WriteLine(errorOutputFormat, cannotInstallReason);
+                this.tracer.RelatedError($"{nameof(this.TryRunProductUpgrade)}: Upgrade is not installable. {cannotInstallReason}");
+                return false;
+            }
+
+            if (!this.TryLoadUpgradeRing(out ring, out error))
+            {
+                this.tracer.RelatedError($"{nameof(this.TryRunProductUpgrade)}: Could not load upgrade ring. {error}");
+                this.ReportInfoToConsole(GVFSConstants.UpgradeVerbMessages.InvalidRingConsoleAlert);

```

Comment:
The opposite of what I had reported before. I now configured an invalid ring, and this is telling me no ring is set.

```
C:\Repos\VFSForGit\src>gvfs config upgrade.ring Foo

C:\Repos\VFSForGit\src>gvfs upgrade
Upgrade ring is not set. No upgrade check was performed.

ERROR: Invalid upgrade ring `Foo` specified in gvfs config.
To set or change upgrade ring, run `gvfs config upgrade.ring ["Fast"|"Slow"|"None"]` from a command prompt.
```

I think we need to have different behavior for no config vs invalid config. No config behaves like None, it's not an error, though we should tell them how to configure it. An invalid ring is an error, and it shouldn't say that it's not set.

Reply:
Added NoConfig Ring. None and NoConfig will not be treated as errors anymore. No change in behavior for Invalid. Attaching the latest screenshot.

<img width="765" alt="screen shot 2018-10-04 at 3 37 40 pm" src="https://user-images.githubusercontent.com/378580/46505869-74353780-c800-11e8-9b0a-1644f1400eec.png">